File: study/lua/src/app/ui/PopupChooseTactics.lua
--
-- Author: Wangyu
-- Date: 2020-2-19 17:30:09
-- 战法装配选择界面
local PopupBase = require("app.ui.PopupBase")
local PopupChooseTactics = class("PopupChooseTactics", PopupBase)
local PopupChooseTacticsCell = require("app.ui.PopupChooseTacticsCell")
local TypeConvertHelper = require("app.utils.TypeConvertHelper")

function PopupChooseTactics:ctor()
	self._commonNodeBk = nil --弹框背景
	local resource = {
		file = Path.getCSB("PopupChooseTactics", "common"),
		binding = {
			
		}
	}
	PopupChooseTactics.super.ctor(self, resource)
end

function PopupChooseTactics:onCreate()
	self._callBack = nil
	self._tacticsIds = {}
	self._commonNodeBk:addCloseEventListener(handler(self, self._onButtonClose))
end

function PopupChooseTactics:onEnter()
	if self._callBack then
		self:refresh()
	end
end

function PopupChooseTactics:onExit()
	
end

function PopupChooseTactics:setTitle(title)
	self._commonNodeBk:setTitle(title)
end

function PopupChooseTactics:refresh()
	self:updateUI(self._pos, self._slot, self._callBack)
end

function PopupChooseTactics:updateUI(pos, slot, callBack)
    self._pos = pos
    self._slot = slot
    self._callBack = callBack
    self._tacticsIds = G_UserData:getTactics():getTacticsListByPos(pos, slot)


	self._count = math.ceil(#self._tacticsIds / 3)

	if self._count==0 then
		self._listView:setVisible(false)
		local emptyType = TypeConvertHelper.TYPE_TACTICS
		self._fileNodeEmpty:updateView(emptyType)
		self._fileNodeEmpty:setButtonString(Lang.get("tactics_choose_empty_button"))
		self._fileNodeEmpty:setVisible(true)
	else
		self._listView:setVisible(true)
		self._listView:setTemplate(PopupChooseTacticsCell)
		self._listView:setCallback(handler(self, self._onItemUpdate), handler(self, self._onItemSelected))
		self._listView:setCustomCallback(handler(self, self._onItemTouch))
		self._listView:resize(self._count)
		self._fileNodeEmpty:setVisible(false)
	end
end

function PopupChooseTactics:_onItemUpdate(item, index)
    local index = index * 3
    local dataList = {}

    for i=1,3 do
        local tacticsId = self._tacticsIds[index + i]
        dataList[i] = tacticsId
    end

	item:update(self._pos, self._slot, dataList, true)
end

function PopupChooseTactics:_onItemSelected(item, index)
	
end

function PopupChooseTactics:_onItemTouch(index, t)
	local tacticsId = self._tacticsIds[index * 3 + t]

	if self._callBack then
		self._callBack(tacticsId)
	end

	self:close()
end

function PopupChooseTactics:_onButtonClose()
	self:close()
end

return PopupChooseTactics

File: study/ts/script/ui/popup/PopupChooseTactics.ts
import { G_UserData } from "../../init";
import { Lang } from "../../lang/Lang";
import { handler } from "../../utils/handler";
import { TypeConvertHelper } from "../../utils/TypeConvertHelper";
import CommonCustomListViewEx from "../component/CommonCustomListViewEx";
import CommonEmptyTipNode from "../component/CommonEmptyTipNode";
import CommonNormalLargePop from "../component/CommonNormalLargePop";
import PopupBase from "../PopupBase";

const { ccclass, property } = cc._decorator;
@ccclass
export default class PopupChooseTactics extends PopupBase {
    @property({
        type: CommonNormalLargePop,
        visible: true
    }) _commonNodeBk: CommonNormalLargePop = null;
    private _callBack: any;
    private _tacticsIds: any[];
    private _pos: any;
    private _slot: any;
    private _count: number;
    @property({
        type: CommonCustomListViewEx,
        visible: true
    }) _listView: CommonCustomListViewEx = null;
    @property({
        type: CommonEmptyTipNode,
        visible: true
    }) _fileNodeEmpty: CommonEmptyTipNode = null;

    @property(cc.Prefab)
    popupChooseTacticsCell: cc.Prefab = null;
    ctor() {
        this._commonNodeBk = null;
    }
    onCreate() {
        this._commonNodeBk.addCloseEventListener(handler(this, this._onButtonClose));
    }
    onEnter() {
        if (this._callBack) {
            this.refresh();
        }
    }
    onExit() {
    }
    setTitle(title) {
        this._commonNodeBk.setTitle(title);
    }
    refresh() {
        this.updateUI(this._pos, this._slot, this._callBack);
    }
    updateUI(pos, slot, callBack) {
        this._pos = pos;
        this._slot = slot;
        this._callBack = callBack;
        this._tacticsIds = G_UserData.getTactics().getTacticsListByPos(pos, slot);
        this._count = Math.ceil(this._tacticsIds.length / 3);
        if (this._count == 0) {
            this._listView.setVisible(false);
            var emptyType = TypeConvertHelper.TYPE_TACTICS;
            this._fileNodeEmpty.updateView(emptyType);
            this._fileNodeEmpty.setButtonString(Lang.get('tactics_choose_empty_button'));
            this._fileNodeEmpty.node.active = (true);
        } else {
            this._listView.setVisible(true);
            this._listView.setTemplate(this.popupChooseTacticsCell);
            this._listView.setCallback(handler(this, this._onItemUpdate), handler(this, this._onItemSelected));
            this._listView.setCustomCallback(handler(this, this._onItemTouch));
            this._listView.resize(this._count);
            this._fileNodeEmpty.node.active = (false);
        }
    }
    _onItemUpdate(item, index) {
        var index2 = index * 3;
        var dataList = {};
        for (var i = 1; i <= 3; i++) {
            var tacticsId = this._tacticsIds[index2 + i - 1];
            dataList[i] = tacticsId;
        }
        item.updateUI(this._pos, this._slot, dataList, true);
    }
    _onItemSelected(item, index) {
    }
    _onItemTouch(index, t) {
        var tacticsId = this._tacticsIds[index * 3 + t - 1];
        if (this._callBack) {
            this._callBack(tacticsId);
        }
        this.close();
    }
    _onButtonClose() {
        this.close();
    }
}